### PR TITLE
Configure lifecycle rules to automatically clear out Chompy uploads.

### PR DESCRIPTION
**Pull request by DFurnes**
### What's this PR do?

This pull request configures an S3 [lifecycle rule](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lifecycle-mgmt.html) to automatically delete files uploaded to the `temporary/` directory in Northstar's S3 bucket.

### How should this be reviewed?

👀

### Any background context you want to provide?

This allows us to ensure that these files don't pile up over time! 📚

### Relevant tickets

References [Pivotal #]().

### Checklist

- [ ] This PR has been added to the relevant Pivotal card.

### CODE REVIEW

File: components/s3_bucket/main.tf [hunk 1]
```diff
@@ -27,6 +27,12 @@ variable "roles" {
   default     = []
 }
 
+variable "temporary_paths" {
+  description = "Automatically delete files in the following paths after 14 days."
+  type        = list(string)
+  default     = []
```
Comment (DFurnes) [inline components/s3_bucket/main.tf:33]: This says that, by default, we don't have any temporary paths. If we set this variable, it has to be an array of strings.

File: components/s3_bucket/main.tf [hunk 2]
```diff
@@ -62,6 +68,26 @@ resource "aws_s3_bucket" "bucket" {
     enabled = var.versioning || var.replication_target != null
   }
 
+  dynamic "lifecycle_rule" {
+    for_each = var.temporary_paths
+
+    content {
+      id      = md5(lifecycle_rule.value)
```
Comment (DFurnes) [inline components/s3_bucket/main.tf:75]: Hmm, I'd added this because (as far as I remembered) this argument was required in the past! When I went to link to that documentation, though, it says it's optional! I'm going to try omitting and see how it goes.

File: components/s3_bucket/main.tf [hunk 3]
```diff
@@ -62,6 +68,26 @@ resource "aws_s3_bucket" "bucket" {
     enabled = var.versioning || var.replication_target != null
   }
 
+  dynamic "lifecycle_rule" {
```
Comment (DFurnes) [inline components/s3_bucket/main.tf:71]: Because of how this `aws_s3_bucket` resource was built by Hashicorp/AWS, we have to configure this lifecycle rule as a dynamic "property" on the `aws_s3_bucket` (rather than as it's own optional resource that we could point to the bucket it's being attached to, like we do with a lot of other things).

Terraform's dynamic blocks are described in more detail [here](https://www.terraform.io/docs/language/expressions/dynamic-blocks.html).

File: components/s3_bucket/main.tf [hunk 4]
```diff
@@ -62,6 +68,26 @@ resource "aws_s3_bucket" "bucket" {
...
+
+      # Delete files in this path after two weeks:
+      expiration {
+        days = 14
+      }
+
+      # If versioned, delete non-current versions at this path after 3 months:
+      noncurrent_version_expiration {
+        days = 90
+      }
```
Comment (DFurnes) [inline components/s3_bucket/main.tf:87]: Since we have versioning enabled on this production bucket, we'd end up with a lot of "deleted" version markers inside this `temporary/` directory. I figured it made sense to clean these up 90 days after the file itself is deleted. (We'll continue to retain other old versions indefinitely for safe-keeping).

File: components/s3_bucket/main.tf [hunk 5]
```diff
@@ -62,6 +68,26 @@ resource "aws_s3_bucket" "bucket" {
...
+      id      = md5(lifecycle_rule.value)
+      prefix  = lifecycle_rule.value
+      enabled = true
+
+      # Delete files in this path after two weeks:
+      expiration {
+        days = 14
+      }
+
+      # If versioned, delete non-current versions at this path after 3 months:
```
Comment (katiecrane) [inline components/s3_bucket/main.tf:84]: What does this mean?

File: components/s3_bucket/main.tf [hunk 6]
```diff
@@ -62,6 +68,26 @@ resource "aws_s3_bucket" "bucket" {
     enabled = var.versioning || var.replication_target != null
   }
 
+  dynamic "lifecycle_rule" {
+    for_each = var.temporary_paths
+
+    content {
+      id      = md5(lifecycle_rule.value)
+      prefix  = lifecycle_rule.value
```
Comment (katiecrane) [inline components/s3_bucket/main.tf:75]: What does the `prefix` do? Where does this value come from? Sorry still not super used to some of this syntax!

File: components/s3_bucket/main.tf (hunk 5 above)
Comment (DFurnes) [reply components/s3_bucket/main.tf:83]: We have versioning enabled on our production bucket, so when we delete a file (or overwrite it, or whatnot) S3 keeps historical copies in case we need to roll back. This gives us extra peace of mind that we never lose user data from a bug! However, it'd also mean we'd keep these temporary files indefinitely unless we add a rule to remove those old versions too.

File: components/s3_bucket/main.tf (hunk 2 above)
Comment (DFurnes) [reply components/s3_bucket/main.tf]: I made this change in d675657 (including removing other `id` values that should no longer be necessary).

File: components/s3_bucket/main.tf (hunk 4 above)
Comment (katiecrane) [reply components/s3_bucket/main.tf:86]: Ah missed this comment on my first pass!

File: components/s3_bucket/main.tf (hunk 6 above)
Comment (DFurnes) [reply components/s3_bucket/main.tf:75]: No worries, it's a little hard to follow!! The `prefix` argument "filters" which objects in the S3 bucket this rule applies to (so by default, it'd apply to everything & this argument limits it to only things that match a given file/folder prefix). The `lifecycle_rule.value` is the "iterator" for this dynamic block & comes from the `var.lifecycle_rule` input variable that we're iterating over.

Comment (DFurnes) [reply components/s3_bucket/main.tf:75]: So, in this case for Northstar, we're iterating over `["temporary/"]` and so we'd loop a single time where `lifecycle_rule.value` would be `"temporary/"` (and thus we'd only enforce these automatic deletion rules for files that were inside that folder).

Comment (katiecrane) [reply components/s3_bucket/main.tf:75]: So in the case of Northstar this limits to stuff in `temporary/`?

Comment (DFurnes) [reply components/s3_bucket/main.tf:75]: Yup! Here's what this turns into (from the plan) after Terraform does all it's fancy looping:

```hcl
lifecycle_rule {
  enabled = true
  prefix  = "temporary/"

  expiration {
    days = 14
  }

  noncurrent_version_expiration {
    days = 90
  }
}
```

Comment (DFurnes) [reply components/s3_bucket/main.tf:75]: Ideally we'd be able to just create this as a standalone `aws_s3_lifecycle_rule` resource in the Northstar module that looked exactly like that & deal with any looping funny-business to build this into the `s3_bucket` module but alas! 🙃

Comment (katiecrane) [reply components/s3_bucket/main.tf:75]: SO HELPFUL!! Thanks for spelling it out much much appreciated! I follow now!

**Review by katiecrane — APPROVED** [review]
Good from me! Thanks for teaching me!

File: components/s3_bucket/main.tf (hunk 2 above)
Comment (DFurnes) [reply components/s3_bucket/main.tf]: Yay, that worked!! It automatically generated a `tf-s3-lifecycle-20210517211330135300000001` ID for this lifecycle rule when it created it. That makes these a lot easier to work with. 😌
